docs/SemanticARC.md

```diff
@@ -128,8 +145,41 @@ In order to simplify this, we will make the following changes:
    * @unowned @safe
    * @forwarding
 
+[EE] We really need a good definition of all these conventions here.
+I strongly believe that the local/global notation would help a lot for a good formal definition of the conventions.
+It's hard for the compiler to reason anything about global RCs. But local RCs are easily statically computable and for me, semantic ARC is all about dealing with local RCs.
+
+So we could define:
+
+@owned operand: local RC must be 1 and is decremented to 0 by the instruction, potentially destroying the object.
+@owned result: Returns the object with a local RC of 1
+@guaranteed operand: local RC must be 1 and is not modified by the instruction
+@unowned operand: local RC must be 0 and is not modified by the instruction.
+@unowned result: Returns the object with a local RC of 0
+@forwarding: Does not change the local RC which can be either 0 or 1
+
+Note that the local RC can only be 0 or 1. It doesn't make sense to have a local RC > 1.
+
+ad @unowned: what do you mean with @unsafe and @safe
+
+ad SILArguments: given the definitions above, what is @owned for a SILArgument? In other words, do we need @owned for SILArgument in addition to @guaranteed and @unowned?
+
+ad @guaranteed: There is a potential alternative definition of @guaranteed for function parameters: the local + global RC must be at least 1. In case the optimizer can figure out that a function does not bring down the global RC to 0, there is no need to have a local RC of 1 for a @guaranteed parameter. So maybe it's better to use a different name for the pure local +1RC convention.
+
+ad @forwarding: I think we don't need this in SIL syntax. It's the default if not @owned, @guaranteed, etc.
+
+
 @forwarding is a new convention that we add to reduce the amount of extra instructions needed to implement this scheme. @forwarding is a special convention intended for instructions that forward RC Identity that for simplictuy will be restricted to forwarding the convention of their def instruction to all of the uses of that instruction. Of course, for forwarding instructions with multiple inputs, we require that all of the inputs have the same convention.
 
+[EE] Do we really need this restriction? What about this example?
+
+bb0(%0 : @guaranteed %X, %1 : @owned %Y):
+  %s = struct $Foo(%0 : %X, %1 : Y)
+  // ... some code
+  %s1 = struct_extract %s, #x
+  %s2 = struct_extract %s, #y
+  %a = apply %f(%s1, %s2) : $@convention(thin) (@guaranteed X, @owned Y)
+
```

Comment:
I feel that we have a disagreement regarding forwarding instructions (please apologize if I misunderstood your proposal in this respect)

Let me talk about ARC "producers" (~ defs) and "consumers" (~ uses).
This is what the ARC verifier checks: does a producer+consumer pair "match". Now the question is what is a producer and what is a consumer.

Rather than considering every SSA def as producer and every use (= operand) as consumer, I would say that only RC-roots are producers, but not any forwarding instruction.
The same for consumers: only side-effect instructions like apply, etc. which have @owned or @guaranteed operands can be consumers, but not operands of forwarding instructions.

In fact, all our side-effect-free instructions, like aggregate and value-extract instructions, are forwarding.
The nice thing about those instructions is that they don't impose a control dependeny.
This means they don't "execute" at a specific program point, although - due to our data-structures - they are located at a specific point in the instruction list.
But we can freely move them around (just maintaining def-use dominance), clone, CSE them. If you would treat them as first-class ARC producers/consumers this important property gets lost.
We would also have the problem of dealing with value-extract instructions which "partially consume" an aggregate (as we discussed in the meeting).

Instead of considering them as consuming and producing ARC values we can view those instructions as nothing more than a "description" of how to get to the RC root.

Example (in SIL pseudocode), assuming StructX has only a single field x:

bb0(%0 : @owned $StructX):
  %1 = struct_extract %0, #x
  apply %f1(@guaranteed %1)
  apply %f2(@owned %1)
  return

Following code is semantically equivalent to the code above and it is perfectly legal.
It would be illegal if struct_extract would be considered as consuming and producing ARC values.

bb0(%0 : @owned $StructX):
  %1 = struct_extract %0, #x   // no matching @owned consumer
  apply %f1(@guaranteed %1)
  %2 = struct_extract %0, #x
  apply %f2(@owned %2)
  return

What does this mean for the ARC verifier? We solve the data-flow for each RC root instead for each SSA value. We also need to explode aggregates, like we do in the data-flow in RLE and DSE.
Still, the verifier would be pretty simple.

I really believe that treating forwarding instructions not as ARC consumers/producers keeps the SIL simple and would be a much less invasive change than otherwise.

Now back to the original question of why would we restrict forwarding instructions to have consistent operand conventions?
This restriction is simply not needed. In fact, imposing this restriction would be more complicated than to not have it.